tensorpipe/channel/cuda_basic/channel_impl.cc

```diff
@@ -30,165 +31,319 @@ ChannelImpl::ChannelImpl(
-        impl.onTempBufferReadyForSend();
+        op.done = true;
+        impl.onSendOpDone();
       }));
+}
 
-  callback(Error::kSuccess);
+void ChannelImpl::sendChunkThroughCpuChannel(Operation op) {
+  TP_VLOG(6) << "Channel " << id_ << " is sending chunk #" << op.chunkId
+             << " of " << op.numChunks << " for buffer #" << op.sequenceNumber
+             << " through CPU channel";
+  CpuBuffer cpuBuffer{op.tmpBuffer.get(), op.length};
+  // FIXME: Avoid copying the op twice.
+  cpuChannel_->send(
+      cpuBuffer,
+      callbackWrapper_([op](ChannelImpl& impl, std::string descriptor) {
+        if (impl.error_) {
+          return;
+        }
+        impl.sendChunkDescriptor(std::move(op), std::move(descriptor));
+      }),
+      callbackWrapper_([op](ChannelImpl& impl) {
+        TP_VLOG(6) << "Channel " << impl.id_ << " is done sending chunk #"
+                   << op.chunkId << " of " << op.numChunks << " for buffer #"
+                   << op.sequenceNumber << " through CPU channel";
+      }));
 }
 
-void ChannelImpl::onTempBufferReadyForSend() {
+void ChannelImpl::sendChunkDescriptor(Operation op, std::string descriptor) {
+  auto nopHolderOut = std::make_shared<NopHolder<std::string>>();
+  nopHolderOut->getObject() = std::move(descriptor);
+  TP_VLOG(6) << "Channel " << id_ << " is sending descriptor for chunk #"
+             << op.chunkId << " of " << op.numChunks << " for buffer #"
+             << op.sequenceNumber;
+  connection_->write(
+      *nopHolderOut,
+      callbackWrapper_([op{std::move(op)}, nopHolderOut](ChannelImpl& impl) {
+        TP_VLOG(6) << "Channel " << impl.id_
+                   << " is done sending descriptor for chunk #" << op.chunkId
+                   << " of " << op.numChunks << " for buffer #"
+                   << op.sequenceNumber;
+      }));
+}
+
+void ChannelImpl::onSendOpDone() {
   while (!sendOperations_.empty()) {
     auto& op = sendOperations_.front();
-    if (!op.ready) {
+    if (!op.done) {
       break;
     }
-
-    if (error_) {
-      op.descriptorCallback(error_, std::string());
-    } else {
-      CpuBuffer cpuBuffer{op.tmpBuffer.get(), op.length};
-      // Keep tmpBuffer alive until cpuChannel_ is done sending it over.
-      // TODO: This could be a lazy callback wrapper.
-      auto callback = callbackWrapper_(
-          [sequenceNumber{op.sequenceNumber},
-           tmpBuffer{std::move(op.tmpBuffer)}](ChannelImpl& impl) {
-            TP_VLOG(5) << "Channel " << impl.id_ << " is done sending buffer #"
-                       << sequenceNumber << " through CPU channel";
-          });
-      TP_VLOG(6) << "Channel " << id_ << " is sending buffer #"
-                 << op.sequenceNumber << " through CPU channel";
-      cpuChannel_->send(
-          cpuBuffer, std::move(op.descriptorCallback), std::move(callback));
+    if (op.callback) {
+      op.callback(error_);
+    }
+    if (!error_) {
+      sendChunkThroughCpuChannel(std::move(op));
     }
```

Comment:
One thing I'm not sure is what happens if the context is being closed once the send callback fires, but before the tensor is actually sent through the CPU channel?